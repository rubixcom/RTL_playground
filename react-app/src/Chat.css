@keyframes appear {
    from {opacity: 0; transform: translateX(-1000%) scale(0.0)}
    to {opacity: 100;}
}

.MessageBubble {
    background: white;
    display: block;
    margin: 20px;
    max-width: 50%;
    padding: 20px;
    border-radius: 10px;
    min-height: 50px;
    animation-name: appear;
    animation-duration: 1s;
    text-align: left;
}

.UserPic {
    height: 50px;
    width: 50px;
    object-fit: cover;
    float: left;
    margin-right: 10px;

    line-height:50px;
    text-align:center;
    background:white;

}

.ChatStream {
    height: 100vh;
    overflow: scroll;
}

.TimeStamp {
    font-weight: 800;
    display: block;
    text-align: left;
}

.Message {
    display: inline;
    text-align: left;
}