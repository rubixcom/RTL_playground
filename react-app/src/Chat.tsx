import React from 'react';
import './Chat.css';
// @ts-ignore
import ScrollToBottom from 'react-scroll-to-bottom';
import moment from 'moment'
import { thisTypeAnnotation } from '@babel/types';
import { async } from 'q';

interface Message {
    picURL: string;
    fromName: string;
    timestamp: number;
    text: string;
}

interface Props { 
}
interface State {
    messages: Message[];
}

interface MessageBubbleProps {
    message: Message;
}

const MessageBubble = ({ message }: MessageBubbleProps) => (
    <div data-testid="MessageBubble" className="MessageBubble">
        <img data-testid="userPic" className="UserPic" alt="Cat pic" src={message.picURL}></img>
        <span data-testid="timeStamp" className="TimeStamp">{moment(message.timestamp).format('hh:mm a')}</span>
        <span data-testid="message" className="Message">{message.text}</span>
    </div>
)

class Chat extends React.Component<Props, State> {
    constructor(props: React.Props<Props>) {
        super(props as any);
        this.state = {
            messages: []
        }
    }

    componentDidMount() {
        const that = this;
        const fetcher = () => {
            that.fetchNewMessages().then(() => { }).catch(() => { });
            setTimeout(fetcher, 1000);
        }
        fetcher();
    }

    async fetchNewMessages() {
        const data = await
            fetch('http://localhost:3001/messages', {
                method: 'GET',
                headers: { 'Content-Type': 'application/json' }
            });
        const messages = await data.json();
        let messageArray = this.state.messages.concat(messages);
        this.setState({ messages: messageArray });
    }
    
  render() {
    const { messages } = this.state;
    return (
      <div className="App">
            <ScrollToBottom className="ChatStream">
                {messages.map((message: Message, index: number) =>
                    <MessageBubble key={index} message={message} />)}
        </ScrollToBottom>
      </div>
    );
  }
}

export default Chat;
