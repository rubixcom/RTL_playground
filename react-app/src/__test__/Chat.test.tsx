import React from 'react';
import { render, cleanup, waitForElement, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import Chat from '../Chat';
import { FetchMock } from 'jest-fetch-mock';

afterEach(cleanup);

beforeEach(() => {
    (fetch as FetchMock).resetMocks()
})
  
const messageMock = {
    picURL: 'http://www.example.com/cat.png',
    fromUser: 'Cat',
    timestamp: 100,
    text: 'This is cat talking'
};

it("matches snapshot", () => {
  const { asFragment } = render(<Chat />);
  expect(asFragment()).toMatchSnapshot();
});

it('displays no messages initially', () => {
    (fetch as FetchMock).mockResponseOnce(JSON.stringify([]));
    const { queryByTestId } = render(<Chat />);
    const messageBubble = queryByTestId('MessageBubble');
    expect(messageBubble).toBeNull();
});

it('displays one message', async () => {
    (fetch as FetchMock).mockResponse(JSON.stringify([messageMock]))

    const { getByTestId } = render(<Chat />);
    const firstMessage = await waitForElement(() =>
        getByTestId('MessageBubble')
    );
    expect(firstMessage.children.length).toBeGreaterThan(0);

    const picture = within(firstMessage).getByTestId('userPic') as HTMLImageElement;
    expect(picture.src).toBe(messageMock.picURL);
    expect(within(firstMessage).getByTestId('timeStamp').textContent).toBe('10:00 am');
    expect(within(firstMessage).getByTestId('message').textContent).toBe(messageMock.text);
});