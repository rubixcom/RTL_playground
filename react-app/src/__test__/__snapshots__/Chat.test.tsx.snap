// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="css-13azwyo ChatStream"
    >
      <div
        class="css-y1c0xs"
      />
    </div>
  </div>
</DocumentFragment>
`;
